client/app/components/ParameterMappingInput.jsx

```diff
@@ -253,28 +359,81 @@ export class ParameterMappingListInput extends React.Component {
-
-          return (
-            <div key={mapping.name} className={(index === 0 ? '' : ' m-t-15')}>
-              <ParameterMappingInput
-                mapping={mapping}
-                existingParamNames={existingParamsNames}
-                onChange={newMapping => this.updateParamMapping(mapping, newMapping)}
-                clientConfig={clientConfig}
-                Query={Query}
-              />
-            </div>
-          );
-        })}
+      <div ref={this.wrapperRef} className="paramMappingList">
+        <Table
+          dataSource={dataSource}
+          size="middle"
+          pagination={false}
+          rowKey={(record, idx) => `row${idx}`}
+        >
+          <Table.Column
+            title="Edit"
+            dataIndex="mapping"
+            key="edit"
+            render={(mapping) => {
+              const existingParamsNames = existingParams
+                .filter(({ type }) => type === mapping.param.type) // exclude mismatching param types
+                .map(({ name }) => name); // keep names only
+
+              return (
+                <EditMapping
+                  mapping={mapping}
+                  existingParamNames={existingParamsNames}
+                  onChange={(oldMapping, newMapping) => this.updateParamMapping(oldMapping, newMapping)}
+                  getContainerElement={() => this.wrapperRef.current}
+                  clientConfig={clientConfig}
+                  Query={Query}
+                />
+              );
+            }}
+          />
+          <Table.Column
+            title="Title"
+            dataIndex="mapping"
+            key="title"
+            render={mapping => mapping.title || mapping.param.title}
+          />
+          <Table.Column
+            title="Keyword"
+            dataIndex="mapping"
+            key="keyword"
+            className="keyword"
+            render={mapping => <code>{`{{ ${mapping.name} }}`}</code>}
+          />
+          <Table.Column
+            title="Default Value"
+            dataIndex="mapping"
+            key="value"
+            render={mapping => this.constructor.getDefaultValue(mapping)}
+          />
+          <Table.Column
+            title="Value Source"
+            dataIndex="mapping"
+            key="source"
+            render={(mapping) => {
+              switch (mapping.type) {
+                case MappingType.DashboardAddNew:
+                case MappingType.DashboardMapToExisting:
+                  return (
+                    <Fragment>
+                      Dashboard parameter{' '}
+                      <Tag className="tag">{mapping.mapTo}</Tag>
+                    </Fragment>
+                  );
+                case MappingType.WidgetLevel:
+                  return 'Widget parameter';
+                case MappingType.StaticValue:
+                  return 'Static value';
+                default:
+                  return ''; // won't happen (typescript-ftw)
```

Comment:
Typescript knows if your [switch case is exhaustive](https://www.typescriptlang.org/play/#src=enum%20ENUM%20%7B%0D%0A%20%20%20%20enum1%2C%0D%0A%20%20%20%20enum2%2C%0D%0A%7D%3B%0D%0A%0D%0Afunction%20foo(arg%3A%20ENUM)%3A%20string%20%7B%0D%0A%20%20%20%20switch(arg)%20%7B%0D%0A%20%20%20%20%20%20%20%20case%20ENUM.enum1%3A%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20'enum1'%3B%0D%0A%20%20%20%20%20%20%20%20case%20ENUM.enum2%3A%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20'enum2'%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%20), so it's safe to ditch `default`.

Reply:
Add `// no default` [comment](https://eslint.org/docs/rules/default-case) instead

Reply:
But then if a new `MappingType` is introduced, this would yield a runtime error. So I'd rather leave it.

(What static typing allows is to omit `default` safely - by getting a compile error `Not all code paths return a value.` if not all cases have been handled.)